app/src/main/java/ru/ifmo/md/extratask1/ImageLoadTask.java

```diff
@@ -0,0 +1,43 @@
+package ru.ifmo.md.extratask1;
+
+import android.content.Context;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+
+public class ImageLoadTask implements ImageTask {
+    Context ctx;
+    private ImageCacher cacher;
+    private String url;
+
+    public ImageLoadTask(ImageCacher cacher, Context ctx, String url) {
+        this.cacher = cacher;
+        this.ctx = ctx;
+        this.url = url;
+    }
+
+    @Override
+    public void run() {
+        String resId = new File(url).getName();
+        if (cacher != null && !cacher.isAvailable(resId)) {
+            try {
+                InputStream is = (InputStream) new URL(url).getContent();
+                cacher.set(resId, is);
+            } catch (IOException e) {
+                // just don't store anything in cache
+            }
+        }
+    }
+
+    public void afterRun() {
+        if (ctx instanceof ImagePreview) {
+            ImagePreview activity = (ImagePreview) ctx;
+            activity.onImageLoad();
+        } else if (ctx instanceof ResultsList) {
```

Comment:
I think it's better implemented via introducing interface with onImageLoad() methods and making both ImagePreview and ResultsList implement this interface, or via [Handler mechanism](http://androidcookbook.com/Recipe.seam;jsessionid=C7CD6C9C9CFBD490DCE7EA024F774381?recipeId=921)